lxd/acme.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+
+	"github.com/gorilla/mux"
+
+	"github.com/lxc/lxd/lxd/acme"
+	"github.com/lxc/lxd/lxd/cluster"
+	"github.com/lxc/lxd/lxd/db/operationtype"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/lxd/task"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+var apiACME = []APIEndpoint{
+	acmeChallengeCmd,
+}
+
+var acmeChallengeCmd = APIEndpoint{
+	Path: ".well-known/acme-challenge/{token}",
+
+	Get: APIEndpointAction{Handler: acmeProvideChallenge, AllowUntrusted: true},
+}
+
+func acmeProvideChallenge(d *Daemon, r *http.Request) response.Response {
+	token, err := url.PathUnescape(mux.Vars(r)["token"])
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// If we're clustered, forwared the request to the leader if necessary.
+	// That is because only the leader knows the token, and any other node will return 404.
+	clustered, err := cluster.Enabled(d.db.Node)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if clustered {
+		leader, err := d.gateway.LeaderAddress()
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		// This gives me the correct value
+		clusterAddress := d.State().LocalConfig.ClusterAddress()
+
+		if clusterAddress != "" && clusterAddress != leader {
+			// Forward the request to the leader
+			client, err := cluster.Connect(leader, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return response.SmartError(err)
+			}
+
+			return response.ForwardedResponse(client, r)
+		}
+	}
+
+	if d.http01Provider == nil || d.http01Provider.Token() != token {
+		return response.NotFound(nil)
+	}
+
+	return response.ManualResponse(func(w http.ResponseWriter) error {
+		w.Header().Set("Content-Type", "text/plain")
+
+		_, err := w.Write([]byte(d.http01Provider.KeyAuth()))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func autoRenewServerCert(ctx context.Context, d *Daemon) error {
```

Comment:
Can we call this `autoRenewClusterCertificate` please

Reply:
I'd rather call it `autoRenewCertificate` as it handles both clusters and single servers.

Reply:
Good idea.